src/EditorFeatures/CSharp/MisplacedUsingDirectives/MisplacedUsingDirectivesCodeFixProvider.cs

```diff
@@ -0,0 +1,328 @@
+                        index = i + 1;
+                    }
+                }
+
+                if (index < triviaList.Count)
+                {
+                    yield return triviaList.TakeRange(index, triviaList.Count - 1);
+                }
+            }
+        }
+
+        private static TSyntaxNode EnsureLeadingBlankLineBeforeFirstMember<TSyntaxNode>(TSyntaxNode node) where TSyntaxNode : SyntaxNode
+        {
+            var members = GetMembers(node);
+            if (members.Count == 0)
+            {
+                return node;
+            }
+
+            var firstMember = members.First();
+            var firstMemberTrivia = firstMember.GetLeadingTrivia();
+
+            // If the first member already contains a leading new line then, this will already break up the usings from these members.
+            if (firstMemberTrivia.Count > 0 && firstMemberTrivia.First().IsKind(SyntaxKind.EndOfLineTrivia))
+            {
+                return node;
+            }
+
+            var newFirstMember = firstMember.WithLeadingTrivia(firstMemberTrivia.Insert(0, SyntaxFactory.CarriageReturnLineFeed));
+            return node.ReplaceNode(firstMember, newFirstMember);
+        }
+
+        private static (AddImportPlacement placement, bool preferPreservation) DeterminePlacement(CompilationUnitSyntax compilationUnit, OptionSet options)
+        {
+            var codeStyleOption = options.GetOption(CSharpCodeStyleOptions.PreferredUsingDirectivePlacement);
+
+            var placement = codeStyleOption.Value;
+            var preferPreservation = codeStyleOption.Notification == NotificationOption.None;
+
+            if (preferPreservation || placement == AddImportPlacement.OutsideNamespace)
+            {
+                return (codeStyleOption.Value, preferPreservation);
+            }
+
+            // Determine if we can safely apply the InsideNamespace preference.
+
+            // Do not offer a code fix when there are multiple namespaces in the source file. When there are
+            // nested namespaces it is not clear if inner usings can be moved outwards without causing 
+            // collisions. Also, when moving usings inwards it is complex to determine which namespaces they
+            // should be moved into.
+
+            // Only move using declarations inside the namespace when
+            // - There are no global attributes
+            // - There are no type definitions outside of the single top level namespace
+            // - There is only a single namespace declared at the top level
+            var forcePreservation = compilationUnit.AttributeLists.Any()
+                || compilationUnit.Members.Count > 1
+                || CountNamespaces(compilationUnit.Members) != 1;
+
+            return (AddImportPlacement.InsideNamespace, forcePreservation);
+        }
+
+        private static int CountNamespaces(SyntaxList<MemberDeclarationSyntax> members)
```

Comment:
nit: Since we don't need the actual count, the method can be changed so it returns immediately when more than 1 namespace is encountered